blog/_posts/2017-12-05-ml&pl.md

```diff
@@ -0,0 +1,107 @@
+
+<div style="text-align:center">
+<a href="https://xkcd.com/1838/">
+<img height="350px" src="https://imgs.xkcd.com/comics/machine_learning_2x.png"/>
+</a>
+</div>
+<div class="desc">
+  Software Engineering 2.0? (via XKCD)
+</div>
+
+A downside to any new language is that it require a new library ecosystem, as only code written for the new runtime benefits from it. For example, rather than reusing the Python ecosystem, the TensorFlow developers need to rewrite libraries for tasks like [image processing](https://www.tensorflow.org/api_guides/python/image) and [file IO](https://www.tensorflow.org/api_docs/python/tf/TextLineReader) in the graph language, throwing out the vast effort behind projects like SciPy. This may well be the only way forward, but ML practitioners should not split from the wider numerical and HPC community. An ideal ML ecosystem is an ideal numerical one, and vice versa, and collaboration between these communities will multiply everyone’s efforts.
+
+We expect to see these developments coming from several angles. Graph IRs and formats like [XLA](https://www.tensorflow.org/performance/xla/), [ONNX](https://github.com/onnx/onnx) and [NNVM](https://github.com/dmlc/nnvm) are becoming ever more sophisticated, and one could imagine them adding more features and even surface syntax to become fully-fledged programming languages[^chris]. TensorFlow’s XLA started a push towards special-purpose compiler stacks that now includes [TVM](http://tvmlang.org/), [DLVM](http://dlvm.org/), [myelin](https://github.com/google/sling/tree/master/myelin), and other ongoing work. Meanwhile, projects like the [PyTorch JIT](https://github.com/pytorch/pytorch/tree/master/torch/csrc/jit), [Gluon](https://mxnet.incubator.apache.org/api/python/gluon.html) and [Tangent](https://github.com/google/tangent) are heroic efforts to make Python itself a better modelling language, in spite of enormous challenges. Having just argued that ML is a numerical programming languages problem, we in the Julia community feel that it is an excellent substrate for experimenting with these language-level issues, and will continue to push the boundaries with projects like [Knet](https://github.com/denizyuret/Knet.jl), [Flux](https://fluxml.github.io/), [Cassette](https://github.com/jrevels/Cassette.jl), [CUDAnative](https://github.com/JuliaGPU/CUDAnative.jl), [DataFlow.jl](https://github.com/MikeInnes/DataFlow.jl), and more.
+
+[^chris]: Google Brain’s recent hiring of Chris Lattner, a well-known language designer and compiler developer, is an interesting development on this point.
```

Comment:
I don't love this part since that is just speculation; we don't know what he's working on.

Reply:
Would you love it if it was labelled as speculative?

Reply:
I think the point we're trying to make here is that ML people are becoming interested in PL expertise (aligning with our argument), not that anyone might be working on a particular project. But this isn't clear and I'll try and reword it tomorrow.

Reply:
I think a rewording that makes it clear as speculative, but evidence that ML teams are beginning to take PL more seriously - all the arguments in the article, and Google hiring a high profile PL person - lends credibility to the story. It all adds up. We are not trying to speculate anything.